pkg/fpgacontroller/fpgacontroller.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fpgacontroller implements reconiciling controller for
+// AcceleratorFunction and FpgaRegion objects used in the FPGA admission
+// webhook.
+package fpgacontroller
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	fpgav2 "github.com/intel/intel-device-plugins-for-kubernetes/pkg/apis/fpga.intel.com/v2"
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/fpgacontroller/patcher"
+)
+
+// +kubebuilder:rbac:groups=fpga.intel.com,resources=acceleratorfunctions;fpgaregions,verbs=get;list;watch
+
+// AcceleratorFunctionReconciler reconciles AcceleratorFunction objects.
+type AcceleratorFunctionReconciler struct {
+	client.Client
+	Log            logr.Logger
+	Scheme         *runtime.Scheme
+	PatcherManager *patcher.PatcherManager
+}
+
+// Reconcile reconciles updates for AcceleratorFunction objects.
+func (r *AcceleratorFunctionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("af", req.NamespacedName)
+
+	p := r.PatcherManager.GetPatcher(req.NamespacedName.Namespace)
+	var af fpgav2.AcceleratorFunction
+	if err := r.Get(ctx, req.NamespacedName, &af); err != nil {
+		if apierrors.IsNotFound(err) {
+			p.RemoveAf(req.NamespacedName.Name)
+			log.V(4).Info("removed from patcher")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch AcceleratorFunction object")
+		return ctrl.Result{}, err
+	}
+
+	log.V(4).Info("received", "AcceleratorFunction", af)
+	return ctrl.Result{}, p.AddAf(&af)
+}
+
+// SetupWithManager sets up the controller.
+func (r *AcceleratorFunctionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fpgav2.AcceleratorFunction{}).
+		Complete(r)
+}
+
+// FpgaRegionReconciler reconciles AcceleratorFunction objects.
+type FpgaRegionReconciler struct {
+	client.Client
+	Log            logr.Logger
+	Scheme         *runtime.Scheme
+	PatcherManager *patcher.PatcherManager
+}
+
+// Reconcile reconciles updates for FpgaRegion objects.
+func (r *FpgaRegionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("af", req.NamespacedName)
```

Comment:
Do we need `r.Log` if the main logger is available from `log.FromContext`?

Reply:
`r.Log` is given its own namespace (`ctrl.Log.WithName("controllers").WithName("FpgaRegion")`). We can drop them if they don't add value.

Basically here I'm following the code examples for kubebuilder:
https://github.com/kubernetes-sigs/kubebuilder/blob/3c4eb4f92a8d9fad1ba2d411343bb42e37526c4c/docs/book/src/cronjob-tutorial/testdata/project/controllers/cronjob_controller.go#L90

Reply:
It looks the same can be done with the latest changes like this: https://github.com/kubernetes-sigs/controller-runtime/blob/master/examples/crd/main.go#L46

Reply:
I see, things are changing...

This change to the Reconciler interface was accepted to the master branch two weeks ago and hasn't been released yet. I'd wait for the next release and then adopt this new interface.

Reply:
> I see, things are changing...

Yes. I also noticed the change is pretty recent but it was documented by the link you shared and that's how I spotted it.